stackslib/src/chainstate/nakamoto/tenure.rs

```diff
@@ -0,0 +1,1107 @@
+        let sql = "SELECT * FROM nakamoto_block_headers WHERE consensus_hash = ?1 ORDER BY block_height ASC LIMIT 1";
+        query_row_panic(chainstate_conn, sql, &[&consensus_hash], || {
+            "FATAL: multiple rows for the same consensus hash".to_string()
+        })
+        .map_err(ChainstateError::DBError)
+    }
+
+    /// Get the last block header in a Nakamoto tenure
+    pub fn get_nakamoto_tenure_finish_block_header(
+        chainstate_conn: &Connection,
+        consensus_hash: &ConsensusHash,
+    ) -> Result<Option<StacksHeaderInfo>, ChainstateError> {
+        let sql = "SELECT * FROM nakamoto_block_headers WHERE consensus_hash = ?1 ORDER BY block_height DESC LIMIT 1";
+        query_row_panic(chainstate_conn, sql, &[&consensus_hash], || {
+            "FATAL: multiple rows for the same consensus hash".to_string()
+        })
+        .map_err(ChainstateError::DBError)
+    }
+
+    /// Get the number of blocks in a tenure.
+    /// Only works for Nakamoto blocks, not Stacks epoch2 blocks.
+    /// Returns 0 if the consensus hash is not found.
+    pub fn get_nakamoto_tenure_length(
+        chainstate_conn: &Connection,
+        consensus_hash: &ConsensusHash,
+    ) -> Result<u32, ChainstateError> {
+        let sql = "SELECT IFNULL(COUNT(block_hash),0) FROM nakamoto_block_headers WHERE consensus_hash = ?1";
+        let count_i64 = query_int(chainstate_conn, sql, &[&consensus_hash])?;
+        let count: u32 = count_i64
+            .try_into()
+            .expect("FATAL: too many blocks in tenure");
+        Ok(count)
+    }
+
+    /// Get the highest coinbase height processed.
+    /// Returns Ok(Some(coinbase_height)) if we have processed at least one tenure
+    /// Returns Ok(None) if we have not yet processed a Nakamoto tenure
+    /// Returns Err(..) on database errors
+    pub fn get_highest_nakamoto_coinbase_height(
+        conn: &Connection,
+        max: u64,
+    ) -> Result<Option<u64>, ChainstateError> {
+        match conn
+            .query_row(
+                "SELECT IFNULL(MAX(coinbase_height), 0) FROM nakamoto_tenures WHERE coinbase_height < ?1",
+                &[&u64_to_sql(max)?],
+                |row| Ok(u64::from_row(row).expect("Expected u64 in database")),
+            )
+            .optional()?
+        {
+            Some(height_i64) => {
+                if height_i64 == 0 {
+                    // this never happens, so it's None
+                    Ok(None)
+                } else {
+                    Ok(Some(
+                        height_i64.try_into().map_err(|_| DBError::ParseError)?,
+                    ))
+                }
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Get the highest processed tenure on the canonical sortition history.
+    pub fn get_highest_nakamoto_tenure<SH: SortitionHandle>(
+        conn: &Connection,
+        sort_handle: &mut SH,
+    ) -> Result<Option<NakamotoTenure>, ChainstateError> {
```

Comment:
Would `SortitionDB::get_canonical_block_header()` be useful here?

It seems something like this would be equivalent:

```rust
let stacks_tip = SortitionDB::get_canonical_block_header();
let nakamoto_tenure_at_tip = query("SELECT * FROM nakamoto_tenures WHERE tenure_id_consensus_hash = ?", stacks_tip.consensus_hash);
return nakamoto_tenure_at_tip
```

Reply:
It would, but then the resulting function would be harder to test.  Can this wait until after this PR?

Reply:
I'd strongly prefer that this happen in this PR. Otherwise, this method contains what is essentially a parallel implementation of "get the canonical stacks tip", and we should prevent duplication of code not only for its own sake, but because parallel implementations not only confuse the codebase (what's the canonical way to obtain the chain tip?), but also introduce the possibility of difference between the implementations, which would be a problem.

Reply:
Okay, this is done.